sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ConvertToLocalRelation.scala

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * Converts empty query stage to empty `LocalRelation`
+ */
+object ConvertToLocalRelation extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case l @ LogicalQueryStage(_, stage: QueryStageExec) if stage.resultOption.get().isDefined &&
```

Comment:
Sorry I made the wrong decision. This may change the output partitioning and is not always safe/beneficial (we may add extra shuffles in the planning phase later).

Looking at the optimizations for empty local relations, some of them are likely beneficial and we can always do: eliminate join, aggregate, limit, repartition, sort, generate

Some may not that beneficial and we shouldn't do: simplify union, eliminate project/filter/sample.

My new idea:
1. Create a new rule `PropagateEmptyRelationBasic`, which deals with local relaion only, and is not very beneficial (eliminate project, filter, etc), and runs in the normal optimizer only
2. Create a new rule `PropagateEmptyRelationAdvanced`, which deals with both local relation and query stage, and is very beneficial (eliminate join, aggregate, etc.), and runs in both normal and AQE optimizer

The old `EliminateUnnecessaryJoin` and `PropagateEmptyRelation` rules should be removed and merged into the new rules.

`PropagateEmptyRelationAdvanced` may not be able to access QueryStageExec which is in sql/core. We can let this rule take `checkRowCount` functiton as a parameter.

Reply:
I see the issue. If we worry about the `LocalRelation` output partitioning, we can just mark `LocalRelationExec` output partitioning as `SinglePartition` to avoid extra shuffle. But it doesn't work with optimization like non empty left semi/anti elimintaion.

> we can always do: eliminate join, aggregate

Not sure we can always do this if we don't want to introduce extra shuffle. And the issue it has already existed in current `EliminateUnnecessaryJoin`, like this plan
```
Aggregate (same key with join)
  Join Inner
    LocalRelation
    xxx
```

An another idea, if we plan to support extra shuffle later and don't expect introduce shuffle at AQE optimzier side, then is it better to check the physical plan `requiredChildDistribution` ? We can only allow one node which has a valid `requiredChildDistribution` (not `UnspecifiedDistribution`) in one query stage, and skip optimize if one query stage has two or more valid `requiredChildDistribution` nodes. Thus we can run `PropagateEmptyRelation` safely.

Reply:
If we turn a broadcast stage into local relation without changing other plan parts, seems we will broadcast the local relation again. `SinglePartition` can't help here.

Reply:
If it's hard to avoid introduce shuffle at AQE optimizer, how about add extra shuffle check between AQE optimizer and stage preparation ? Then it won't affect the extra shuffle in stage preparation.